Add influxdb-cpp-rest/1.0.2

d-led opened the pull request:
This PR adds the influxdb-cpp-rest package version 1.0.2 to Conan Center.

## Package Details
- **Name**: influxdb-cpp-rest
- **Version**: 1.0.2
- **License**: MPL-2.0
- **Homepage**: https://github.com/d-led/influxdb-cpp-rest

## Features
- C++20 client library for InfluxDB
- Includes both a C++ static library (influxdb-cpp-rest) and a C wrapper shared library (influx-c-rest)
- Supports InfluxDB 1.x (tested with v1.8+)
- Synchronous and asynchronous insertion APIs
- Query API support
- Basic authentication support

## Dependencies
- cpprestsdk/2.10.19
- rxcpp/4.1.1
- catch2/3.11.0 (test requirement only)

## Compiler Requirements
- C++20 or later
- GCC 13+ (for std::format support), Clang 14+, MSVC 19.29+ (VS 2019 16.10+)

## Changes from 1.0.1
- Added compiler version validation to prevent builds on unsupported compilers
- Updated minimum GCC requirement from 11+ to 13+ (std::format requires GCC 13+)

Automatically created by GitHub Actions workflow on tag push.

jcar87 commented on recipes/influxdb-cpp-rest/1.0.1/conanfile.py:
```diff
@@ -0,0 +1,78 @@
+        tc = CMakeToolchain(self)
+        # Disable tests and demo for packaging
+        tc.variables["BUILD_TESTING"] = False
+        tc.variables["BUILD_DEMO"] = False
```
```suggestion
        tc.cache_variables["BUILD_TESTING"] = False
        tc.cache_variables["BUILD_DEMO"] = False
```

`cache_variables` are provided via `-D` flags as users would, whereas `variables` are in-line defined in the loaded `conan_toolchain.cmake` - so we probably want cache variables here

d-led commented:
@franramirez688 added a small compile smoke test. Does that work? https://github.com/conan-io/conan-center-index/pull/28789/commits/9544c8cbbfaad9318484417ffaccaaae0f5fe57c
Shall I remove the outputs?

franramirez688 commented:
Hi @d-led,

The CI is failing for [Linux](https://github.com/conan-io/conan-center-index/pull/28789/checks?check_run_id=54591282188) and [macOS](https://github.com/conan-io/conan-center-index/pull/28789/checks?check_run_id=54591254779) at least. Could you have a look at the errors? They are compilation ones that may need a look into.

d-led commented:
ping

jcar87 commented on recipes/influxdb-cpp-rest/all/conanfile.py:
```diff
@@ -0,0 +1,119 @@
+                raise ConanInvalidConfiguration(
+                    f"influxdb-cpp-rest requires Clang 14+ for std::format support. "
+                    f"Current version: {self.settings.compiler.version}"
+                )
```
This should be simplified as per above message, but also this should also cover the condition `compiler.libcxx=libc++`

franramirez688 commented on recipes/influxdb-cpp-rest/all/conanfile.py:
```diff
@@ -52,6 +53,38 @@ def source(self):
+                raise ConanInvalidConfiguration(
+                    f"influxdb-cpp-rest requires MSVC 19.29+ (VS 2019 16.10+) for std::format support. "
+                    f"Current version: {self.settings.compiler.version}"
+                )
```
Should it not be Clang `15` and Apple Clang `14`?
Also, we could refactor these lines as follows:

```python
        minimum_compiler_versions = {"gcc": 13, "clang": 15, "apple-clang": 14, "msvc": 192}
        minimum_version = minimum_compiler_versions.get(str(self.settings.compiler))
        if minimum_version and Version(self.settings.compiler.version) < minimum_version:
            raise ConanInvalidConfiguration(f"Requires {self.settings.compiler} >= {minimum_version} for std::format support")
```

franramirez688 commented:
> can we restart the build with the new recipe?

Hi @d-led, we restarted the CI when you pushed the latest commit https://github.com/conan-io/conan-center-index/pull/28789/commits/78d0edf8828e78abf90fc32fac0a283969888425, and macos failed as you can see here https://github.com/conan-io/conan-center-index/pull/28789/checks?check_run_id=56531021227

d-led commented:
Thanks for the patience — addressed the CI failures and review comments:

- **macOS build error**: the library used `std::format` with a runtime format string, which on libc++ targets < macOS 13.3 instantiates the floating-point formatter (needs `std::to_chars(float)`, unavailable before 13.3). I removed `std::format` from the library entirely and released **v1.0.3** (it was the only C++20-only feature, so the GCC/Clang version gates are gone too).
- **Recipe**: changed to `package_type = "static-library"` and dropped the `shared` option — the C++ core is `add_library(... STATIC)` upstream, so a shared option didn't do anything. This also resolves the `fpic_management` hook warning.
- **`package_info`**: added `includedirs = ["include/influxdb-cpp-rest"]` — headers install under that subdir, so the test package couldn't include `<influxdb_line.h>` before.
- **test_package**: fixed a typo (`kvpkey_value_pairs` → `key_value_pairs`) and made the test minimal.

Local `conan create` passes end-to-end (build + test_package).

d-led commented:
Hi @franramirez688 @jcar87. sorry for the delays on my side. I've tried yet again to address the comments and the macOS CI failure (the library no longer requires std::format, released as v1.0.3). Could you please trigger a new CI run? Happy to fix anything just to make it.

franramirez688 reviewed (changes requested):
Hi @d-led,

Thanks a lot for solving the error 🙏 Now, I have a couple of questions regarding this latest version:

* OpenSSL is required upstream but it's not reflected here, so we need to add it to the `requirements`. It's working as it's a transitive dep, but we need to tell consumers that it depends on it.
* I saw that we're installing two libs. One of them is modelled by the package_info and the other one not. The idea would be to have two components, one for each lib:
  * static-only: `libinfluxdb-cpp-rest` (already modelled and installed)
  * shared-only: `libinflux-c-rest` (not modelled yet but installed)

d-led commented:
Hi @franramirez688, thanks for the quick reply! Indeed, the library supports two targets - as a c++ lib, and as a C wrapper usable from e.g. other languages or from C (given the run-time). I'll fix the recipe to name OpenSSL explicitly. Indeed, it's neede for the cpprestsdk client. I'll make some fixes in a while

d-led commented:
@franramirez688 pushed the fixes

- Added `openssl/[>=1.1 <4]` to `requirements()`
- Modeled both libraries as components: `cpp-rest` (static `influxdb-cpp-rest`) and `c-rest` (shared `influx-c-rest`), each with its own CMake target and include dirs
- rxcpp is header only and if I understand correctly, `transitive_headers` is needed for working usage of `cpprestsdk`
- Added a C-wrapper smoke test in the test package

it worked locally. looking forward to CI

franramirez688 commented on recipes/influxdb-cpp-rest/all/conanfile.py:
```diff
@@ -0,0 +1,82 @@
+        # C wrapper (shared)
+        self.cpp_info.components["c-rest"].set_property("cmake_target_name", "influxdb-cpp-rest::influx-c-rest")
+        self.cpp_info.components["c-rest"].libs = ["influx-c-rest"]
+        self.cpp_info.components["c-rest"].includedirs = ["include/influx-c-rest"]
```
we need to define the requirements here too:

```suggestion
        self.cpp_info.components["c-rest"].includedirs = ["include/influx-c-rest"]
        self.cpp_info.components["c-rest"].requires = ["cpp-rest"]
```

franramirez688 commented on recipes/influxdb-cpp-rest/all/test_package/test_c_rest.cpp:
```diff
@@ -0,0 +1,20 @@
+    const char* result = influx_c_rest_lines_get(lines);
+
+    const bool ok = result != nullptr
+        && std::strstr(result, "measurement") != nullptr
+        && std::strstr(result, "value=42i") != nullptr;
+
+    influx_c_rest_lines_destroy(lines);
+    influx_c_rest_key_value_pairs_destroy(values);
+
+    return ok ? 0 : 1;
+}
```
Could we simplify this? Something like:

```cpp
  #include <influx_c_rest_lines.h>

  int main()
  {
      auto* values = influx_c_rest_key_value_pairs_new();
      influx_c_rest_key_value_pairs_destroy(values);
      return 0;
  }
```

d-led replied:
@franramirez688 thanks again for the suggestions and the review 🙏. Pushed the simplification. 

P.S. sorry for being a bit slow. Fell out of the c++ world for many years, thus have forgotten the sharp tools 😄

d-led replied:
Should be done already

franramirez688 reviewed (commented):
@d-led Looking great! Just a couple of comments 🙏

franramirez688 commented on recipes/influxdb-cpp-rest/all/conanfile.py:
```diff
@@ -0,0 +1,83 @@
+        self.cpp_info.components["c-rest"].set_property("cmake_target_name", "influxdb-cpp-rest::influx-c-rest")
+        self.cpp_info.components["c-rest"].libs = ["influx-c-rest"]
+        self.cpp_info.components["c-rest"].includedirs = ["include/influx-c-rest"]
+        self.cpp_info.components["c-rest"].requires = ["cpp-rest"]
```
@d-led, we would need to set this to avoid the error of consuming the shared library:

```suggestion
        self.cpp_info.components["c-rest"].requires = ["cpp-rest"]
        self.cpp_info.components["c-rest"].type = "shared-library"
        # Adding the "bin" folder to work around this https://github.com/conan-io/conan/issues/20222
        self.runenv_info.append_path("PATH", os.path.join(self.package_folder, "bin"))
```

This is actually an open issue: https://github.com/conan-io/conan/issues/20222

UPDATE: this line `self.cpp_info.components["c-rest"].type = "shared-library"` does not solve the problem, but it's more explicit, at least

franramirez688 replied:
With that UPDATE, we also need to update this:

```python
required_conan_version = ">=2.9"
```

That `type` attribute was introduced in Conan 2.9

franramirez688 commented on recipes/influxdb-cpp-rest/all/conanfile.py:
```diff
@@ -0,0 +1,86 @@
+from conan.tools.cmake import CMake, CMakeDeps, CMakeToolchain, cmake_layout
+from conan.tools.files import copy, get, replace_in_file
+
+required_conan_version = ">=2.1"
```
Because of https://github.com/conan-io/conan-center-index/pull/28789/changes#r3828961472

```suggestion
required_conan_version = ">=2.9"
```

uilianries reviewed (approved):
LGTM

Hook warning is a false-positive, the build-type is mapped by the component already.

franramirez688 commented:
Thanks @d-led 👏

d-led commented:
oof, thanks a lot, @franramirez688, @jcar87  @uilianries !
